File: tests/python/test_creators_schema.py
#!/usr/bin/env python3
"""契约测试：creators.json schema 校验。
type/role 是自由标签，只校验必填字段和格式。"""
import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent.parent
JSON_FILE = ROOT / "creators.json"


def validate():
    errors = []

    if not JSON_FILE.exists():
        errors.append(f"MISSING: {JSON_FILE}")
        return errors

    data = json.loads(JSON_FILE.read_text("utf-8"))
    if not isinstance(data, list) or len(data) == 0:
        errors.append("SCHEMA: must be a non-empty array")
        return errors

    names = set()
    for i, creator in enumerate(data):
        if not isinstance(creator, dict):
            continue

        prefix = f"[{i}] {creator.get('name', '?')}"

        name = creator.get("name", "")
        if not name or not isinstance(name, str):
            errors.append(f"{prefix}: 'name' must be non-empty string")
        elif name in names:
            pass  # duplicate names are allowed (multiple source entries)
        names.add(name)

        for field in ["desc", "type", "role"]:
            val = creator.get(field, "")
            if val and not isinstance(val, str):
                errors.append(f"{prefix}: '{field}' must be string")

    return errors


def main():
    errors = validate()
    if errors:
        sys.stdout.buffer.write(f"FAILED: {len(errors)} schema violation(s) in {JSON_FILE.name}\n\n".encode("utf-8"))
        for e in errors:
            sys.stdout.buffer.write(f"  {e}\n".encode("utf-8"))
        sys.exit(1)
    else:
        sys.stdout.buffer.write(b"OK: all creators passed schema checks\n")


if __name__ == "__main__":
    main()

File: tests/test_creators_schema.mjs
#!/usr/bin/env node
/**
 * 契约测试：creators.json schema 校验。
 * type/role 是自由标签，只校验必填字段和格式。
 * 由 tests/python/test_creators_schema.py 迁移（2026-08-03），校验逻辑逐点保真。
 */
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const JSON_FILE = path.join(ROOT, 'creators.json');

function validate() {
  const errors = [];

  if (!fs.existsSync(JSON_FILE)) {
    errors.push(`MISSING: ${JSON_FILE}`);
    return errors;
  }

  let data;
  try {
    data = JSON.parse(fs.readFileSync(JSON_FILE, 'utf-8'));
  } catch (e) {
    errors.push(`SYNTAX: creators.json 解析失败: ${e.message}`);
    return errors;
  }

  if (!Array.isArray(data) || data.length === 0) {
    errors.push('SCHEMA: must be a non-empty array');
    return errors;
  }

  const names = new Set();
  for (let i = 0; i < data.length; i++) {
    const creator = data[i];
    if (typeof creator !== 'object' || creator === null) continue;

    const prefix = `[${i}] ${creator?.name ?? '?'}`;

    const name = creator?.name ?? '';
    if (!name || typeof name !== 'string') {
      errors.push(`${prefix}: 'name' must be non-empty string`);
    }
    // duplicate names are allowed (multiple source entries)
    names.add(name);

    for (const field of ['desc', 'type', 'role']) {
      const val = creator?.[field] ?? '';
      if (val && typeof val !== 'string') {
        errors.push(`${prefix}: '${field}' must be string`);
      }
    }
  }

  return errors;
}

const errors = validate();
if (errors.length) {
  console.error(`FAILED: ${errors.length} schema violation(s) in ${path.basename(JSON_FILE)}\n`);
  for (const e of errors) console.error(`  ${e}`);
  process.exit(1);
} else {
  console.log('OK: all creators passed schema checks');
}
